From 386821dda5b58621aeaabd0bba1e270b817d3de9 Mon Sep 17 00:00:00 2001
From: Shreyansh Jain <shreyansh.jain@nxp.com>
Date: Mon, 14 Jan 2019 18:38:33 +0530
Subject: [PATCH 484/741] vfio/fsl-mc: Reset container on last reference
 release

commit c0c2141106240d61e7b9d7c80284711d54f6ae91 from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

Current code have a bug where container is reset on every
reference release. While we need to reset the dprc container
on last reference release.
This patch fixes this by having reset container on last reference
release.

Signed-off-by: Shreyansh Jain <shreyansh.jain@nxp.com>
Signed-off-by: Bharat Bhushan <bharat.bhushan@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/vfio/fsl-mc/vfio_fsl_mc.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/drivers/vfio/fsl-mc/vfio_fsl_mc.c b/drivers/vfio/fsl-mc/vfio_fsl_mc.c
index 682405cf36b8..c7d63dc151cc 100644
--- a/drivers/vfio/fsl-mc/vfio_fsl_mc.c
+++ b/drivers/vfio/fsl-mc/vfio_fsl_mc.c
@@ -127,11 +127,12 @@ static void vfio_fsl_mc_release(void *device_data)
 	if (!(--vdev->refcnt)) {
 		vfio_fsl_mc_regions_cleanup(vdev);
 		vfio_fsl_mc_irqs_cleanup(vdev);
-	}
 
-	if (strcmp(mc_dev->obj_desc.type, "dprc") == 0)
-		dprc_reset_container(mc_dev->mc_io, 0, mc_dev->mc_handle,
-				     mc_dev->obj_desc.id);
+		if (strcmp(mc_dev->obj_desc.type, "dprc") == 0)
+			dprc_reset_container(mc_dev->mc_io, 0,
+					     mc_dev->mc_handle,
+					     mc_dev->obj_desc.id);
+	}
 
 	mutex_unlock(&driver_lock);
 
-- 
2.17.1

